code/game/gamemodes/miniantags/pulsedemon/pulsedemon.dm

```diff
@@ -0,0 +1,823 @@
+		else
+			message = copytext(message, 3)
+
+	message = trim_left(message)
+
+	var/list/message_pieces = list()
+	if(ignore_languages)
+		message_pieces = message_to_multilingual(message)
+	else
+		message_pieces = parse_languages(message)
+
+	// hivemind languages
+	if(istype(message_pieces, /datum/multilingual_say_piece))
+		var/datum/multilingual_say_piece/S = message_pieces
+		S.speaking.broadcast(src, S.message)
+		return TRUE
+
+	if(!LAZYLEN(message_pieces))
+		. = FALSE
+		CRASH("Message failed to generate pieces. [message] - [json_encode(message_pieces)]")
+
+	create_log(SAY_LOG, "[message_mode ? "([message_mode])" : ""] '[message]'")
+
+	playsound(get_turf(src), pick(speech_sounds), 30, TRUE)
+	if(istype(loc, /obj/item/radio))
+		var/obj/item/radio/R = loc
+		name = gen_speech_name()
+		R.talk_into(src, message_pieces, message_mode, verbage = verb)
+		name = real_name
+		return TRUE
+	else if(istype(loc, /obj/machinery/hologram/holopad))
+		var/obj/machinery/hologram/holopad/H = loc
+		name = "[H]"
+		for(var/mob/M in get_mobs_in_view(7, H))
+			M.hear_say(message_pieces, verb, FALSE, src)
+		name = real_name
+		return TRUE
+
+	emote("me", message = "[pick(emote_hear)]")
+	return TRUE
+
+/mob/living/simple_animal/demon/pulse_demon/update_runechat_msg_location()
+	if(istype(loc, /obj/machinery/hologram/holopad))
+		runechat_msg_location = loc.UID()
+	else
+		return ..()
+
+/mob/living/simple_animal/demon/pulse_demon/visible_message(message, self_message, blind_message)
+	// overriden because pulse demon is quite often in non-turf locs, and /mob/visible_message acts differently there
+	for(var/mob/M in get_mobs_in_view(7, src))
+		if(M.see_invisible < invisibility)
+			continue //can't view the invisible
+		var/msg = message
+		if(self_message && M == src)
+			msg = self_message
+		M.show_message(msg, EMOTE_VISIBLE, blind_message, EMOTE_AUDIBLE)
+
+/mob/living/simple_animal/demon/pulse_demon/has_internal_radio_channel_access(mob/user, list/req_one_accesses)
+	return has_access(list(), req_one_accesses, get_all_accesses())
+
+/mob/living/simple_animal/demon/pulse_demon/proc/try_hijack_apc(obj/machinery/power/apc/A, remote = FALSE)
+	// one APC per pulse demon, one pulse demon per APC, no duplicate APCs
+	if(!is_valid_apc(A) || (A in hijacked_apcs) || apc_being_hijacked || A.being_hijacked)
+		return FALSE
+
+	do_hijack_notice(A)
+	apc_being_hijacked = A
+	A.being_hijacked = TRUE
+	A.update_icon()
+	if(do_after(src, hijack_time, target = A))
+		if(is_valid_apc(A))
+			finish_hijack_apc(A, remote)
+		else
+			fail_hijack(A)
+	cleanup_hijack_apc(A)
+
+// note: the station maps supposedly average ~150 APCs, so the upper levels here are certainly possible, also you can manually upgrade the capacity stat
+/mob/living/simple_animal/demon/pulse_demon/proc/calc_maxcharge(n)
+	if(n < 1)
+		return 1000
+	return 20000 * clamp(n, 0, 20) + 500000 * clamp(n - 20, 0, 30) + 1000000 * clamp(n - 50, 0, 50) + 500000000 * max(0, n - 100)
```

Comment:
dear god, I know this wasn't your change gdn so feel free to dismiss this but it would be really nice if there was a hint of documentation here

Reply:
I'll procdef this, looks like an atmos proc lol